### PR TITLE
[v1.2] Add Module Class

### PR DESCRIPTION
**Pull request by rohit0718**
Fixes [#36]

Let's:
1. Add Module, ModuleName, and ModuleCode classes
2. Implement parseModule, parseModuleName, and parseModuleCode methods in ParserUtil
3. Add tests for Module, ModuleName, and ModuleCode
4. Update ParserUtilTest to test  parseModule, parseModuleName, and parseModuleCode

### CODE REVIEW

File: src/main/java/seedu/address/commons/util/AppUtil.java [hunk 1]
```diff
@@ -39,4 +41,21 @@ public static void checkArgument(Boolean condition, String errorMessage) {
             throw new IllegalArgumentException(errorMessage);
         }
     }
+
+    /**
+     * Checks that lists {@code first} and {@code second} have the same number of elements and that
+     * each element in index i in {@code first} is equal (based on the .equals() function) to the element
+     * in index i in {@code second}.
+     */
+    public static <T> boolean areListsEqual(List<T> first, List<T> second) {
```
Comment (bentanjunrong) [inline src/main/java/seedu/address/commons/util/AppUtil.java:50]: What is this equality used for? Looks like it only allows for lists whose elements are in the same order. Not sure if that's the expected behaviour you're going for.

File: src/main/java/seedu/address/model/module/ModuleCode.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+ */
+public class ModuleCode {
+    /*
+     * "Each module of study has a unique module code consisting of a two‐ or three‐letter prefix
+     * that denotes the discipline, and four digits, the first of which indicates the level of
+     * the module (e.g., 1000 indicates a Level 1 module and 2000, a Level 2 module)."
+     * There is also an optional last character representing the course of study.
+     * Reference: https://www.nus.edu.sg/registrar/academic-information-policies/non-graduating/modular-system
+     */
+    public static final String VALIDATION_REGEX = "^([a-zA-Z]{2,3}[0-9]{4}[a-zA-Z]?)$";
```
Comment (bentanjunrong) [inline src/main/java/seedu/address/model/module/ModuleCode.java:18]: Regex looks good

File: src/test/java/seedu/address/logic/parser/ParserUtilTest.java [hunk 3]
```diff
@@ -419,4 +426,63 @@ public void parseLesson_validValues_returnsLesson() throws Exception {
                 "Lecture", "Monday", "09:00", "11:00",
                 Optional.of(VALID_VENUE), Optional.of(VALID_LINK)));
     }
+
+    @Test
+    public void parseModuleCode_null_throwsNullPointerException() {
```
Comment (bentanjunrong) [inline src/test/java/seedu/address/logic/parser/ParserUtilTest.java:431]: Tests look good

File: src/main/java/seedu/address/model/module/Module.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+import java.util.Optional;
+
+import seedu.address.model.module.exam.Exam;
+import seedu.address.model.module.lesson.Lesson;
+
+/**
+ * Represents a Module in the ModBook.
+ * Guarantees: immutable.
+ */
+public class Module {
```
Comment (bentanjunrong) [inline src/main/java/seedu/address/model/module/Module.java:18]: Missing method for get upcoming lesson. Unless the plan is to save it for the next iteration.

**Review by bentanjunrong — CHANGES_REQUESTED** [review]

File: src/main/java/seedu/address/model/module/Module.java (hunk 4 above)
Comment (rohit0718) [reply src/main/java/seedu/address/model/module/Module.java:17]: yup adding later on!

**Pull request merged**